### PR TITLE
Gio/networking server refactor

### PR DESCRIPTION
**Pull request by sinkingsugar**

### CODE REVIEW

File: src/core/shards/network.cpp [hunk 1]
```diff
@@ -254,45 +274,106 @@ struct Server : public NetworkBase {
...
+
+            // keep receiving
+            return do_receive();
+          }
+        });
   }
 
   std::vector<uint8_t> _buffer;
 
+  IterableExposedInfo _sharedCopy;
```
Comment (guusw) [inline src/core/shards/network.cpp:373]: Can you use `ExposedInfo` instead? I want to get rid of this weird IterableArray eventually.

This should be possible:
```cpp
_sharedCopy = ExposedInfo(data.shared);
_sharedCopy.push_back(...);
```

File: src/core/shards/network.cpp [hunk 2]
```diff
@@ -335,7 +457,58 @@ struct Server : public NetworkBase {
...
+
+  static inline Type PeerType{{SHType::Object, {.object = {.vendorId = CoreCC, .typeId = PeerCC}}}};
+  static inline Type PeerObjectType = Type::VariableOf(PeerType);
+
+  static SHTypesInfo outputTypes() { return PeerType; }
+
   NetworkPeer _peer;
+  ShardsVar _blks{};
+
+  static inline ParamsInfo params = ParamsInfo(
```
Comment (guusw) [inline src/core/shards/network.cpp:477]: These need to be in requiredVariables, you can use the `collectRequiredVariables` helper in compose:
```cpp
ExposedInfo _required;
SHTypeInfo compose(const SHInstanceData &data) {
   _required.clear();
   collectRequiredVariables(data.shared, _required, (SHVar&)_addr);
   collectRequiredVariables(data.shared, _required, (SHVar&)_port);
}
```

Comment (guusw) [reply src/core/shards/network.cpp:477]: Check the other shards as well since they have similar parameters.

**Review by guusw — CHANGES_REQUESTED** [review]

File: src/core/shards/network.cpp (hunk 2 above)
Comment (sinkingsugar) [reply src/core/shards/network.cpp:477]: But shardsvar is not an actual variable, just shards no?

File: src/core/shards/network.cpp (hunk 1 above)
Comment (sinkingsugar) [reply src/core/shards/network.cpp:373]: but why? what you suggested would not work, I need a clone of the whole array.
Also what's weird of IterableArray, just a wrapper on top of the simple C array we have? Lets KISS?

Comment (sinkingsugar) [reply src/core/shards/network.cpp:373]: Btw we do this in many other places.

File: src/core/shards/network.cpp (hunk 2 above)
Comment (sinkingsugar) [reply src/core/shards/network.cpp:477]: `CoreInfo::ShardsOrNone` nothing to require

Comment (guusw) [reply src/core/shards/network.cpp:477]: I meant for _addr and _port

Comment (sinkingsugar) [reply src/core/shards/network.cpp:477]: Ah good point

**Pull request merged**